### PR TITLE
Feature/genetic algorithm

### PR DESCRIPTION
**Pull request by StrelokCH**
Prototype of genetic algorithm according to the milestone.

### CODE REVIEW

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using IRuettae.Core.GeneticAlgorithm.Algorithm.Helpers;
+using IRuettae.Core.GeneticAlgorithm.Algorithm.Models;
+
+namespace IRuettae.Core.GeneticAlgorithm.Algorithm
+{
+    public class RecombinationOperation
+    {
+        private double orderBasedCrossoverProbability = 0.5;
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs]: readonly

Comment (Isitar) [reply IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs]: remove 0.5, wird im konstruktor überschrieben

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/MutationOperation.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+using System.Security.Cryptography;
+using IRuettae.Core.GeneticAlgorithm.Algorithm.Helpers;
+using IRuettae.Core.GeneticAlgorithm.Algorithm.Models;
+
+namespace IRuettae.Core.GeneticAlgorithm.Algorithm
+{
+    public class MutationOperation
+    {
+        private const double ProbabilityPositionMutation = 0.5;
+        private const double ProbabilityInversionMutation = 1.0 - ProbabilityPositionMutation;
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/MutationOperation.cs]: never used

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/CostCalculator.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+using IRuettae.Core.Models;
+
+namespace IRuettae.Core.GeneticAlgorithm.Algorithm
+{
+    /// <summary>
+    /// Class to recalculate the cost of Genotypes
+    /// </summary>
+    public class CostCalculator
+    {
+        private OptimizationResult result;
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/CostCalculator.cs]: readonly

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Models/Genotype.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    /// <summary>
+    /// Class to hold a list of alleles
+    /// The size and the alleles should not change,
+    /// but the order of the alleles can change.
+    /// </summary>
+    public class Genotype : List<int>
+    {
+        public int Cost { get; set; }
+
+        public Genotype() : base()
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Models/Genotype.cs]: base() entfernen

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Decoder.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    /// Class to convert a Genotype into a Route[]
+    /// </summary>
+    public class Decoder
+    {
+        private readonly OptimizationInput input;
+        private readonly Dictionary<int, int> alleleToVisitIdMapping;
+        /// <summary>
+        /// Cached list of all santaIds
+        /// </summary>
+        private List<int> santaIds = new List<int>();
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Decoder.cs]: readonly

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/CostCalculator.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        private Decoder decoder;
+
+        public CostCalculator(Decoder decoder, OptimizationResult temporaryResult)
+        {
+            this.decoder = decoder;
+            this.result = temporaryResult;
+        }
+
+        public void RecalculatateCost(IEnumerable<Genotype> population)
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/CostCalculator.cs]: typo

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        //private double probabilityEdgeRecombinationCrossover = 1.0 - probabilityOrderBasedCrossover;
+        private Random random;
+
+        /// <summary>
+        /// Factor to be multiplied with population size
+        /// to get minimum number of positions to retain in OrderBasedCrossover
+        /// </summary>
+        private const double MinumumNumberOfPositionsFactor = 1d / 4d;
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs]: typo

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/MutationOperation.cs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="population"></param>
+        /// <param name="probability">between 0 and 1</param>
+        public void Mutate(List<Genotype> population, double probability)
+        {
+            for (int i = 0; i < population.Count; i++)
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/MutationOperation.cs]: convert to foreach

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RepairOperation.cs [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+        public RepairOperation(OptimizationInput input, Dictionary<int, int> alleleToVisitIdMapping)
+        {
+            if (alleleToVisitIdMapping == null)
+            {
+                throw new ArgumentNullException();
+            }
+
+            this.input = input;
+            needRepair = input.Visits.Where(v => v.IsBreak).Count() > 0;
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RepairOperation.cs]: any

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        /// Factor to be multiplied with population size
+        /// to get maximum number of positions to retain in OrderBasedCrossover
+        /// </summary>
+        private const double MaximumNumberOfPositionsFactor = 3d / 4d;
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="random">not null</param>
+        public RecombinationOperation(Random random, double orderBasedCrossoverProbability)
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs]: documentation for 2. param

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        public OptimizationResult Solve(long timelimitMiliseconds, EventHandler<ProgressReport> progress, EventHandler<string> consoleProgress)
+        {
+            this.progress = progress;
+            this.consoleProgress = consoleProgress;
+
+            var sw = Stopwatch.StartNew();
+            Log("Solving started");
+            Log(new ProgressReport(0.01));
+
+            // adjust timelimit if unlimited
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs]: timeLimit

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Decoder.cs [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+                    r.SantaId = santaIds[routePerDayCounter];
+                    routes.Add(r);
+                    routePerDayCounter++;
+                }
+            }
+            return routes.ToArray();
+        }
+
+        /// <summary>
+        /// Depending on the dataset, artifical santas must be created
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Decoder.cs]: artificial typo

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+            // calculate costs
+            var result = new OptimizationResult()
+            {
+                OptimizationInput = input,
+            };
+            var costCalculator = new CostCalculator(decoder, new SimplifiedOptimizationResult(result));
+            costCalculator.RecalculatateCost(population);
+
+            // Log characteristics of inital population
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs]: typo

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RepairOperation.cs [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+                var breakId = input.Visits.Where(v => v.IsBreak && v.SantaId == santa.Id).First().Id;
+                var breaks = alleleToVisitIdMapping.Where(e => e.Value == breakId).Select(e => e.Key).ToArray();
+                breakMapping.Add(santa.Id, breaks);
+            }
+        }
+
+        /// <summary>
+        /// Repairs the given genotype in place
+        /// </summary>
+        /// <param name="genotype"></param>
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RepairOperation.cs]: wrong param name

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Decoder.cs [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+            };
+            while (currentGenPos < genotype.Count && !PopulationGenerator.IsSeparator(genotype[currentGenPos]))
+            {
+                int visitId = GetVisitId(genotype[currentGenPos]);
+                var previousWaypoint = waypoints.Last();
+                int startTime = previousWaypoint.StartTime;
+                if (waypoints.Count == 1)
+                {
+                    // first real visit
+                    startTime += input.Visits.Where(v => v.Id == visitId).First().WayCostFromHome;
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/Decoder.cs]: lambda first(v=> ...)

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+        /// <param name="report"></param>
+        public void Log(ProgressReport report)
+        {
+            progress?.Invoke(this, report);
+        }
+
+        /// <summary>
+        /// convenience
+        /// </summary>
+        /// <param name="report"></param>
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs]: param name

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        /// <returns></returns>
+        public int GetMinCost(IEnumerable<Genotype> population)
+        {
+            return population.Select(i => i.Cost).Min();
+        }
+
+        public static double StdDev(IEnumerable<double> values)
+        {
+            double ret = 0.0;
+            var count = values.Count();
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/GenAlgSolver.cs]: multiple enumarations, ändern auf readonlylist ?

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+                {
+                    lowestNumberOfNeighbours = numberOfNeighbours;
+                    nextAllele = allele;
+                }
+            }
+
+            if (!nextAllele.HasValue)
+            {
+                // take first allele
+                nextAllele = neighbours.Keys.Where(allele => allele != previousAllele).First();
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs]: first(allele => ...)

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RepairOperation.cs [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+                    }
+                }
+            }
+            return ret;
+        }
+
+        /// <summary>
+        /// Removes the first occurence of the allel
+        /// </summary>
+        /// <param name="genotype"></param>
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RepairOperation.cs]: param name

Comment (Isitar) [reply IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RepairOperation.cs]: add allele to doc

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+        private static Dictionary<int, List<int>> GetNeighbours(Genotype parent1, Genotype parent2)
+        {
+            // allele to neighbours
+            var neighbours = new Dictionary<int, ISet<int>>();
+            foreach (var allele in parent1)
+            {
+                neighbours[allele] = new SortedSet<int>();
+            }
+
+            void addNeightbour(int pos, int posNeighbour)
```
Comment (Isitar) [inline IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs]: rename to AddNeighbour

File: IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs (hunk 1 above)
Comment (StrelokCH) [reply IRuettae/Core/IRuettae.Core.GeneticAlgorithm/Algorithm/RecombinationOperation.cs]: und readonly machen!

**Pull request merged**